### PR TITLE
[FRS-16]Support reporting metrics to prometheus

### PR DESCRIPTION
**Pull request by dangshazi**
## What is the purpose of the change
Support reporting metrics to prometheus

## Brief change log

  - *Support reporting metrics to prometheus*

## Verifying this change

Add Tests:
- `com.alibaba.flink.shuffle.metrics.reporter.PrometheusReporterTest`
- `ConfigurationParserUtilsTest`

This change is already covered by existing tests, such as `com.alibaba.flink.shuffle.metrics.reporter.ReporterSetupTest`

### CODE REVIEW

**Comment by wsry** [comment]
Thanks for opening this PR, I will take a look soon.

**Review by wsry — COMMENTED** [review]
@dangshazi Thanks for the improvement. I have left some comments.

**Comment by wsry** [comment]
@dangshazi BTW, we updated the license header today because we were informed that project which is still not contributed to apache can not use the apache license header. As a result, you also need to update that.

**Comment by dangshazi** [comment]
> , so there may be something I can't figure

I'll add more test

File: shuffle-core/src/main/java/com/alibaba/flink/shuffle/core/config/MetricOptions.java [hunk 1]
```diff
@@ -22,6 +22,30 @@
...
+     * remote-shuffle.metrics.reporters = foo,bar
+     *
+     * remote-shuffle.metrics.reporter.foo.class = org.apache.flink.metrics.reporter.JMXReporter
+     * remote-shuffle.metrics.reporter.foo.interval = 10
+     *
+     * remote-shuffle.metrics.reporter.bar.class = org.apache.flink.metrics.graphite.GraphiteReporter
+     * remote-shuffle.metrics.reporter.bar.port = 1337
+     * }</pre>
+     */
+    public static final ConfigOption<String> REPORTERS_LIST =
```
[the remark this answers is not in the document]
Comment (dangshazi) [reply shuffle-core/src/main/java/com/alibaba/flink/shuffle/core/config/MetricOptions.java:42]: Done

File: shuffle-metrics-reporter/shuffle-metrics-reporter-prometheus/src/main/java/com/alibaba/flink/shuffle/metrics/reporter/PrometheusReporter.java [hunk 2]
```diff
@@ -0,0 +1,153 @@
...
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.util.Iterator;
+import java.util.Map;
+import java.util.concurrent.TimeUnit;
+
+/** {@link MetricManagerReporter} that exports {@link Metric Metrics} via Prometheus. */
+public class PrometheusReporter extends AbstractPrometheusReporter {
```
[the remark this answers is not in the document]
Comment (dangshazi) [reply shuffle-metrics-reporter/shuffle-metrics-reporter-prometheus/src/main/java/com/alibaba/flink/shuffle/metrics/reporter/PrometheusReporter.java:36]: Yes  The Prometheus server will pull metrics from the shuffle cluster actively

I think pull mode of Prometheus is suitable for service with Fix ips

File: shuffle-metrics-reporter/shuffle-metrics-reporter-prometheus/src/main/java/com/alibaba/flink/shuffle/metrics/reporter/PrometheusReporterOptions.java [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+
+    public static final ConfigOption<Duration> INTERVAL =
+            new ConfigOption<Duration>("interval")
+                    .defaultValue(Duration.ofSeconds(10))
+                    .description(
+                            "The interval between reports. This is used as a suffix in an actual reporter config.");
+
+    public static final ConfigOption<String> PORTS =
+            new ConfigOption<String>("ports")
+                    .defaultValue("9349-9359")
```
[the remark this answers is not in the document]
Comment (dangshazi) [reply shuffle-metrics-reporter/shuffle-metrics-reporter-prometheus/src/main/java/com/alibaba/flink/shuffle/metrics/reporter/PrometheusReporterOptions.java:44]: Config the ip range into PrometheusServer configuration file

File: shuffle-core/src/main/java/com/alibaba/flink/shuffle/core/config/MetricOptions.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (dangshazi) [reply shuffle-core/src/main/java/com/alibaba/flink/shuffle/core/config/MetricOptions.java:42]: Removed

File: shuffle-metrics/src/main/java/com/alibaba/flink/shuffle/metrics/reporter/ReporterSetup.java [hunk 4]
```diff
@@ -18,35 +18,73 @@
...
+import java.util.TreeSet;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
 import java.util.stream.Collectors;
-import java.util.stream.Stream;
-
-import static com.alibaba.flink.shuffle.core.config.MetricOptions.METRICS_REPORTER_CLASSES;
 
 /** Encapsulates everything needed for the instantiation and configuration of a metric reporter. */
 public final class ReporterSetup {
```
[the remark this answers is not in the document]
Comment (dangshazi) [reply shuffle-metrics/src/main/java/com/alibaba/flink/shuffle/metrics/reporter/ReporterSetup.java:38]: OK  Removed

File: shuffle-metrics-reporter/shuffle-metrics-reporter-prometheus/src/main/java/com/alibaba/flink/shuffle/metrics/reporter/PrometheusReporter.java [hunk 5]
```diff
@@ -0,0 +1,153 @@
...
+        if ("false".equals(initFlag)) {
+            LOG.info("PrometheusReporter disabled...");
+            return;
+        }
+
+        if (initFlag == null) {
+            System.setProperty(INIT_FLAG, "true");
+            super.start(period, unit);
+        } else {
+            LOG.info("PrometheusReporter has been started...");
```
[the remark this answers is not in the document]
Comment (dangshazi) [reply shuffle-metrics-reporter/shuffle-metrics-reporter-prometheus/src/main/java/com/alibaba/flink/shuffle/metrics/reporter/PrometheusReporter.java:93]: I'm trying to keep consistent with `com.alibaba.metrics.reporter.file.FileMetricManagerReporter`

**Comment by wsry** [comment]
@dangshazi Thanks for the contribution. Merged.

**Pull request closed without merging**